File: 2022/go/main.go
package main

import (
	"adventofcode2022/days"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
)

type arguments struct {
	dayNumber int
	dataDir   string
}

func main() {
	fmt.Println("AoC 2022")
	args := parseArguments()
	inputFilePath := inputFilePathForDay(args)
	solutionPartOne, solutionPartTwo, err := days.RunDay(days.RunDayData{Number: args.dayNumber, InputFilePath: inputFilePath})
	if err != nil {
		panic(err)
	}
	fmt.Printf("Day %d\n\tPart One: %s\n\tPart Two:%s\n", args.dayNumber, solutionPartOne, solutionPartTwo)
}

func inputFilePathForDay(args arguments) string {
	return filepath.Join(args.dataDir, fmt.Sprintf("day%d.txt", args.dayNumber))
}

func parseArguments() arguments {
	if len(os.Args[1:]) != 2 {
		panic("Expecting exactly two arguments.")
	}

	dayNumber, err := strconv.Atoi(os.Args[1])
	if err != nil {
		panic(err)
	}

	if dayNumber < 1 || dayNumber > 24 {
		panic("Only accepting numbers between 1 and 24")
	}

	dataDir := os.Args[2]

	return arguments{dayNumber: dayNumber, dataDir: dataDir}
}

File: 2022/rust/src/main.rs
use std::{env, error, path::PathBuf};

use adventofcode2022::days;

fn main() {
    let args = parse_args().expect("arguments are expected");

    let (part_one, part_two) = days::solve_day(&args.day, &args.file_path());

    println!(
        "Day {}:\n\tPart one: {part_one}\n\tPart two: {part_two}",
        args.day
    );
}

#[derive(Debug)]
struct Arguments {
    data_dir: PathBuf,
    day: u8,
}

impl Arguments {
    fn file_path(&self) -> PathBuf {
        self.data_dir.join(format!("day{}.txt", self.day))
    }
}

fn parse_args() -> Result<Arguments, Box<dyn error::Error>> {
    let args: Vec<String> = env::args().collect();

    let day = match args[1].parse::<u8>()? {
        d @ 1..=24 => d,
        d => return Err(format!("day must be between 1 and 24 but is {d}").into()),
    };

    let data_dir = match args[2].trim() {
        s if !s.is_empty() => PathBuf::from(s),
        _ => return Err("data dir must be passed as argument".into()),
    };

    Ok(Arguments { data_dir, day })
}
